### ISSUE TITLE
llms/generate.py and chat.py  Random Seeds: Default to MLX’s Time-Based RNG or Add a --random-seed Flag

### ISSUE THREAD

**Issue by cavit99** [ml-explore/mlx#1919]
## Description
The `mlx_lm` scripts ([`chat.py`](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/chat.py) and [`generate.py`](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/generate.py)) currently enforce a fixed default seed of `0` for text generation, resulting in identical outputs across runs even when using stochastic sampling (e.g., `--temp 1`). This behavior stems from an explicit call to `mx.random.seed(0)` in both scripts, overriding the natural time-based RNG provided by MLX’s core library in `mlx.core.random`. While this ensures reproducibility, it contrasts with the typical expectation in LLM interfaces—random outputs by default unless a seed is explicitly set—and requires manual workarounds for varied outputs.

I’d like to propose enhancing `mlx_lm` to make random outputs more accessible, either by leveraging MLX’s built-in randomness by default or adding an opt-in flag.

## Current Behavior
- In [`generate.py`](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/generate.py) and [`chat.py`](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/chat.py), the seed is set as follows:
  ```python
  DEFAULT_SEED = 0
  parser.add_argument("--seed", type=int, default=DEFAULT_SEED, help="PRNG seed")
  # ...
  mx.random.seed(args.seed)  # Called unconditionally with default 0
  ```
  This resets the RNG state to `0` unless `--seed` is provided, ensuring deterministic sampling via `mx.random.categorical()` in [`sample_utils.py`](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/sample_utils.py):
  ```python
  def categorical_sampling(logits, temp):
      return mx.random.categorical(logits * (1 / temp))
  ```

- MLX’s core RNG, defined in `mlx.core.random` initializes with a time-based seed:
  ```cpp
  static KeySequence& default_() {
      static KeySequence ks(get_current_time_seed());
      return ks;
  }
  static uint64_t get_current_time_seed() {
      auto now = std::chrono::system_clock::now();
      return std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count();
  }
  ```
  Without the override, this would provide unique random sequences per run.

- _Result_: Outputs are fixed (e.g., same poem for `--prompt "write a poem" --temp 1`) unless users manually specify a unique `--seed` (e.g., `--seed $RANDOM`), which isn’t a built-in option.

## Problem
- This override of `mlx.core.random`’s time-based RNG feels counterintuitive, as most LLM tools (e.g., Hugging Face’s `transformers`, OpenAI SDK) default to random outputs for `temp > 0`, requiring an explicit seed for determinism.
- There’s no native way to leverage MLX’s natural randomness without modifying the code or scripting workarounds, reducing usability for creative or exploratory tasks.

### 1. Default to MLX’s Time-Based RNG (No Seed Set by Default)
- **Change**: Modify [`generate.py`](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/generate.py) and [`chat.py`](https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/chat.py) to skip `mx.random.seed()` unless a seed is explicitly provided. For example:
  ```python
  parser.add_argument("--seed", type=int, default=None, help="PRNG seed (default: use MLX’s time-based RNG)")
  # ...
  if args.seed is not None:
      mx.random.seed(args.seed)  # Only set if provided; otherwise, use MLX’s RNG
  ```
- **Benefit**: Aligns with LLM norms and MLX’s own design, providing random outputs by default while allowing `--seed <int>` for reproducibility.
- **Consideration**: This could break existing workflows expecting identical outputs. The current `type=int` parser would need adjustment to handle `default=None` without type errors.
- **Mitigation**: Introduce in versioning with clear release notes: “Default behavior now uses MLX’s time-based RNG; use `--seed 0` for old behavior.”  I do believe long-term, this would be the most intuitive approach.

### 2. Add a `--random-seed` Flag
- **Change**: Add a flag to opt into MLX’s time-based RNG in both scripts:
  ```python
  parser.add_argument("--random-seed", action="store_true", help="Use MLX’s default time-based RNG instead of a fixed seed")
  # ...
  if args.random_seed:
      pass  # Skip setting seed, rely on mlx.core.random
  else:
      mx.random.seed(args.seed)  # Default to 0 or user-specified
  ```
- **Benefit**: Preserves current behavior (fixed seed) while offering an easy way to get random outputs, avoiding breaking changes.
- **Trade-Off**: Requires users to opt in, keeping the default slightly less aligned with LLM expectations.

## Rationale
- Random outputs enhance usability for creative or casual use cases (e.g., generating varied poems), a common expectation in LLM interfaces.
- MLX’s `mlx.core.random` already provides a time-based seed, so overriding it with `0` feels redundant unless explicitly desired.
- Either option maintains reproducibility via `--seed`, balancing flexibility and control.

## Request for Feedback
Which approach do you think fits `mlx_lm`’s goals best? Option 1 offers a long-term alignment with MLX and LLM norms but requires versioning care. Option 2 is a quick, safe enhancement. I’d be happy to refine this or contribute a PR if there’s interest!
